File: app/src/main/java/com/drdisagree/iconify/xposed/utils/BootLoopProtector.java
package com.drdisagree.iconify.xposed.utils;

/*
 * From AOSPMods
 * https://github.com/siavash79/AOSPMods/blob/canary/app/src/main/java/sh/siava/AOSPMods/modpacks/utils/BootLoopProtector.java
 */

import static com.drdisagree.iconify.config.XPrefs.Xprefs;

import android.annotation.SuppressLint;

import java.util.Calendar;

public class BootLoopProtector {

    public static final String LOAD_TIME_KEY_KEY = "packageLastLoad_";
    public static final String PACKAGE_STRIKE_KEY_KEY = "packageStrike_";

    @SuppressLint("ApplySharedPref")
    public static boolean isBootLooped(String packageName) {
        String loadTimeKey = String.format("%s%s", LOAD_TIME_KEY_KEY, packageName);
        String strikeKey = String.format("%s%s", PACKAGE_STRIKE_KEY_KEY, packageName);
        long currentTime = Calendar.getInstance().getTime().getTime();
        long lastLoadTime = Xprefs.getLong(loadTimeKey, 0);
        int strikeCount = Xprefs.getInt(strikeKey, 0);

        if (currentTime - lastLoadTime > 40000) {
            Xprefs.edit().putLong(loadTimeKey, currentTime).putInt(strikeKey, 0).commit();
        } else if (strikeCount >= 3) {
            return true;
        } else {
            Xprefs.edit().putInt(strikeKey, ++strikeCount).commit();
        }
        return false;
    }

    @SuppressLint("ApplySharedPref")
    public static void resetCounter(String packageName) {
        try {
            String loadTimeKey = String.format("%s%s", LOAD_TIME_KEY_KEY, packageName);
            String strikeKey = String.format("%s%s", PACKAGE_STRIKE_KEY_KEY, packageName);
            long currentTime = Calendar.getInstance().getTime().getTime();

            Xprefs.edit().putLong(loadTimeKey, currentTime).putInt(strikeKey, 0).commit();
        } catch (Throwable ignored) {
        }
    }
}

File: app/src/main/java/com/drdisagree/iconify/xposed/utils/BootLoopProtector.kt
package com.drdisagree.iconify.xposed.utils

import android.annotation.SuppressLint
import com.drdisagree.iconify.xposed.utils.XPrefs.Xprefs
import com.drdisagree.iconify.xposed.utils.XPrefs.XprefsIsInitialized
import java.util.Calendar

object BootLoopProtector {

    const val LOAD_TIME_KEY_KEY = "packageLastLoad_"
    const val PACKAGE_STRIKE_KEY_KEY = "packageStrike_"

    @SuppressLint("ApplySharedPref")
    fun isBootLooped(packageName: String): Boolean {
        if (!XprefsIsInitialized) return false

        val loadTimeKey = "$LOAD_TIME_KEY_KEY$packageName"
        val strikeKey = "$PACKAGE_STRIKE_KEY_KEY$packageName"
        val currentTime = Calendar.getInstance().time.time
        val lastLoadTime = Xprefs.getLong(loadTimeKey, 0)
        val strikeCount = Xprefs.getInt(strikeKey, 0)

        if (currentTime - lastLoadTime > 40000) {
            Xprefs.boundedCommit { editor ->
                editor
                    .putLong(loadTimeKey, currentTime)
                    .putInt(strikeKey, 0)
            }
        } else if (strikeCount >= 3) {
            return true
        } else {
            val newStrikeCount = strikeCount + 1
            Xprefs.boundedCommit { editor ->
                editor.putInt(strikeKey, newStrikeCount)
            }
        }

        return false
    }

    @SuppressLint("ApplySharedPref")
    fun resetCounter(packageName: String) {
        if (!XprefsIsInitialized) return

        try {
            val loadTimeKey = "$LOAD_TIME_KEY_KEY$packageName"
            val strikeKey = "$PACKAGE_STRIKE_KEY_KEY$packageName"
            val currentTime = Calendar.getInstance().time.time

            Xprefs.boundedCommit { editor ->
                editor
                    .putLong(loadTimeKey, currentTime)
                    .putInt(strikeKey, 0)
            }
        } catch (_: Throwable) {
        }
    }
}